libfwupdplugin/fu-smbios.c

```diff
@@ -64,12 +64,71 @@ typedef struct __attribute__((packed)) {
 typedef struct {
 	guint8			 type;
 	guint16			 handle;
-	GBytes			*data;
+	GByteArray		*buf;
 	GPtrArray		*strings;
 } FuSmbiosItem;
 
 G_DEFINE_TYPE (FuSmbios, fu_smbios, G_TYPE_OBJECT)
 
+static void
+fu_smbios_convert_dt_string (FuSmbios *self, guint8 type, guint8 offset,
+			     const gchar *path, const gchar *subpath)
+{
+	FuSmbiosItem *item = g_ptr_array_index (self->items, type);
+	gsize bufsz = 0;
+	g_autofree gchar *fn = g_build_filename (path, subpath, NULL);
+	g_autofree gchar *buf = NULL;
+
+	/* not found */
+	if (!g_file_get_contents (fn, &buf, &bufsz, NULL))
+		return;
+
+	/* add to strtab */
+	g_ptr_array_add (item->strings, g_strndup (buf, bufsz));
+	for (guint i = item->buf->len; i < (guint) offset + 1; i++)
+		fu_byte_array_append_uint8 (item->buf, 0x0);
+	item->buf->data[offset] = item->strings->len;
+}
+
+static gboolean
+fu_smbios_setup_from_path_dt (FuSmbios *self, const gchar *path, GError **error)
+{
+	/* add all four faked structures */
+	for (guint i = 0; i < FU_SMBIOS_STRUCTURE_TYPE_LAST; i++) {
+		FuSmbiosItem *item = g_new0 (FuSmbiosItem, 1);
+		item->type = i;
+		item->buf = g_byte_array_new ();
+		item->strings = g_ptr_array_new_with_free_func (g_free);
+		g_ptr_array_add (self->items, item);
+	}
+
+	/* DMI:Manufacturer */
+	fu_smbios_convert_dt_string (self, FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x04,
+				     path, "vendor");
+
+	/* DMI:Family */
+	fu_smbios_convert_dt_string (self, FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x1a,
+				     path, "model-name");
+
+	/* DMI:ProductName */
+	fu_smbios_convert_dt_string (self, FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x05,
+				     path, "model");
```

Comment:
on a DT system do these also populate `/sys/class/dmi/id`?  If so, maybe it's better to just fallback to that path in fu-smbios.

Reply:
On my qemu system there's no `/sys/class/dmi` whatsoever. @meklort is this populated on actual hardware please?

Reply:
if nothing brings devicetree stuff like this to /sys/class/dmi on real hardware then yes agree this PR looks fine

Reply:
> On my qemu system there's no `/sys/class/dmi` whatsoever. @meklort is this populated on actual hardware please?

Correct, this does not exist on the Talos II or Blackbird systems:

    meklort@talos2 ~/g/b/build ((e8e9e8df…))> ls /sys/class/dmi
    ls: cannot access '/sys/class/dmi': No such file or directory